src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -2269,6 +2239,47 @@ private sealed class DictionaryEnumerator : IDictionaryEnumerator
         }
     }
 
+    internal static class ConcurrentDictionaryTypeProps<T>
```

Comment:
Note that you are paying several hundred bytes in type system data structures per `ConcurrentDictionary` instantiation for this optimization. I do not think it is a problem. `ConcurrentDictionary` instantiations are pretty expensive. Just pointing it out.

Reply:
In my initial attempt I did not have the extra class: https://github.com/dotnet/runtime/pull/77005/commits/6fde9d8242aa640f2f34aa8054f2e6726e509fb1
Of course that does not get as many cases, but I guess there is trade-off as you say.

Reply:
`ConcurrentDictionary<shared_for_generics, not_shared_for_generics>`, like `ConcurrentDictionary<object, int>`.

When the static field is part of the `ConcurrentDictionary`, these cases require generic lookup. The JIT is not able to see that the value depends on `not_shared_for_generics` only.
When the static field is in separate type, these cases do not require generic lookup and the tiered JIT can treat the readonly field as constant.

Reply:
I see. Thanks.